tools: Query teleport-private in addition for latest releases

camscale opened the pull request:
Extend `query-latest` to query releases in `teleport-private` in
addition to the current `teleport` repository so that we can rebuild OCI
images released from `teleport-private`.

If the latest release for a version is in the `teleport-private`
repository, the version number printed will be prefixed with `private-`.

As `teleport-private` is a private repository, the environment variable
`GITHUB_TOKEN` must be set to a token with permissions to read the
releases of the `teleport-private` repository.

When testing locally, this can be done with:

    # may require `gh auth refresh` first
    export GITHUB_TOKEN=$(gh auth token)

assuming you have `gh` set up properly with credentials.

Test-run: https://github.com/gravitational/teleport.e/actions/runs/15601530288
Test-run: https://github.com/gravitational/teleport.e/actions/runs/15601976879

---

Currently the nightly rebuild is disabled (deleted) so changing the
output of this `query-latest` command will not break anything. When I
add back the workflow, I'll adjust it to handle the `private-*` output
properly.

I have created private releases for 15.5.3 and 17.5.2 in teleport-private.
You can see in the first test run of the output of the "Trigger Builds" step
where it would trigger private release builds of 15.5.3 and 17.5.2, but not
16.5.11 for which I have not yet created the private GitHub release.

The second test run shows the rebuild of 15.5.3 from teleport-private, which
was successful. I verified the AWS registry has 15, 15.5 and 15.5.3 on the
same image.

camscale committed:
tools: Query teleport-private in addition for latest releases

Extend `query-latest` to query releases in `teleport-private` in
addition to the current `teleport` repository so that we can rebuild OCI
images released from `teleport-private`.

If the latest release for a version is in the `teleport-private`
repository, the version number printed will be prefixed with `private-`.

As `teleport-private` is a private repository, the environment variable
`GITHUB_TOKEN` must be set to a token with permissions to read the
releases of the `teleport-private` repository.

When testing locally, this can be done with:

    # may require `gh auth refresh` first
    export GITHUB_TOKEN=$(gh auth token)

assuming you have `gh` set up properly with credentials.

r0mant commented on build.assets/tooling/cmd/query-latest/main.go:
```diff
@@ -72,10 +84,26 @@ func parseFlags() (string, error) {
+func getLatestForRepo(ctx context.Context, versionSpec string, gh github.GitHub, repo string) (string, error) {
+	releases, err := gh.ListReleases(ctx, "gravitational", repo)
+	if err != nil {
+		return "", trace.Wrap(err, "couldn't list releases")
```
```suggestion
		return "", trace.Wrap(err, "couldn't list releases for repo %q", repo)
```

fheinecke reviewed (approved):
LGTM provided that this doesn't break v14 builds

camscale commented on build.assets/tooling/cmd/query-latest/main.go:
```diff
@@ -72,10 +84,26 @@ func parseFlags() (string, error) {
+	// Ignore errors on teleport-private - it has sparse releases and may not return any
+	// tags. That's not an error.
+	if err == nil && semver.Compare(publicTag, privateTag) < 0 {
+		return "private-" + privateTag, nil
```
This should work for all versions as the cron workflow runs from master and runs this tool from master. Neither PR addressing the private OCI rebuild needs to be backported.

But the cron workflow only rebuilds v15, v16 and v17, and the query latest too is used only by the oci rebuild cron workflow.

camscale committed:
Add repo name to "couldn't list releases" error message

camscale committed:
Handle no matching result for private separately.

Stop bundling all errors when looking up releases in teleport-private
and ignoring them all. Instead only treat no release matching the search
string as not an error as it is viable that there have not been any
private releases on a branch or minor version. Failure to query github
for private releases should be treated as an error, as not doing so may
mean we ignore a private release that actually is the latest.

camscale commented:
I re-worked this a little bit as I was not comfortable with "Ignore errors on teleport-private", as some errors are real errors and should cause the program to terminate. Only one error is reasonable to be ignored - no matching private releases. We do not want a transient failure to cause a private release not to be considered as it may cause us to republish an older vulnerable release as the latest for a branch.

So I've updated the error handling for this and added some test cases.
